crates/project-model/src/project_json.rs

```diff
@@ -217,6 +260,55 @@ enum EditionData {
     Edition2024,
 }
 
+#[derive(Deserialize, Debug, Clone)]
+pub struct TargetSpecData {
+    manifest_file: PathBuf,
+    target_label: String,
+    target_kind: TargetKindData,
+    runnables: RunnablesData,
+    flycheck_command: Vec<String>,
+}
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct RunnablesData {
+    check: Vec<String>,
+    run: Vec<String>,
+    test: Vec<String>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum TargetKindData {
+    Bin,
+    /// Any kind of Cargo lib crate-type (dylib, rlib, proc-macro, ...).
+    Lib,
+    Example,
+    Test,
+    Bench,
+    BuildScript,
+    Other,
+}
```

Comment:
This ties the cargo workspace architecture into project-json, I'm not sure that this was something @matklad had in mind when giving this alternative format for project descriptions.

Is there a need for things like `BuildScript` and `Example`? `Bin`, `Lib` and `Test` (and `Other` ofc) seem fine as they are rather general.

Reply:
I _think_ https://github.com/rust-lang/rust-analyzer/pull/16135#discussion_r1447221391 answers this, but, yeah good catch of the specific smell here!

Reply:
They're not really needed, no—they were meant to try and make `project-json` more Cargo-like. I only _really_ need  `Bin`, `Lib` and `Test` if this PR doesn't take the approach that matklad suggested in his comment.